process/asyncExecution/queue/blocksQueue.go

```diff
@@ -84,6 +86,74 @@ func (bq *blocksQueue) AddOrReplace(pair HeaderBodyPair) error {
 	return nil
 }
 
+func (bq *blocksQueue) getIndexForNonce(nonce uint64) int {
+	for i, bp := range bq.headerBodyPairs {
+		if bp.Header.GetNonce() == nonce {
+			return i
+		}
+	}
+
+	return -1
+}
+
+func (bq *blocksQueue) getPairsWithHigherNonces(nonce uint64) ([]HeaderBodyPair, int) {
+	pairsToBeRemoved := make([]HeaderBodyPair, 0)
+	firstIndex := len(bq.headerBodyPairs)
+	for i, bp := range bq.headerBodyPairs {
+		if bp.Header.GetNonce() <= nonce {
+			continue
+		}
+
+		if i < firstIndex {
+			firstIndex = i
+		}
+
+		pairsToBeRemoved = append(pairsToBeRemoved, bp)
+	}
+
+	return pairsToBeRemoved, firstIndex
+}
+
+func (bq *blocksQueue) removeAfterNonce(nonce uint64) {
+	pairsToBeRemoved, firstIndex := bq.getPairsWithHigherNonces(nonce)
+	if len(pairsToBeRemoved) == 0 {
+		return
+	}
+
+	bq.notifyEvictedPairs(pairsToBeRemoved)
+
+	bq.headerBodyPairs = bq.headerBodyPairs[:firstIndex]
+	if len(bq.headerBodyPairs) == 0 {
+		bq.lastAddedNonce = nonce
+		return
+	}
+
+	bq.lastAddedNonce = bq.headerBodyPairs[len(bq.headerBodyPairs)-1].Header.GetNonce()
+}
+
+func (bq *blocksQueue) replaceAndRemoveHigherNonces(pair HeaderBodyPair, nonce uint64) error {
+	bq.removeAfterNonce(nonce)
```

Comment:
can we maybe do the remove for the replaced nonce as well? then we get the notify as well the subscribers for the replaced nonce, as it is actually removed, and replaced with another.